payment_gateway_adyen/tests/test_payment.py

```diff
@@ -0,0 +1,217 @@
+        validation = requests.post(validate_url, data={
+            'PaReq': source['paRequest'],
+            'MD': source['md'],
+            'TermUrl': webhook_url,
+            'username': 'user',
+            'password': 'password',
+            'cardNumber': card_number
+        })
+        parser = etree.HTMLParser()
+        tree = etree.parse(StringIO(validation.content.decode('utf-8')),
+                           parser)
+        e = tree.xpath("//input[@name='PaRes']")[0]
+        pa_res = e.values()[2]
+        return pa_res
+
+
+class AdyenScenario(RecordedScenario):
+
+    def test_create_transaction_3d_required_failed(self):
+        self._test_3d('5212345678901234', success=False)
+
+    def test_create_transaction_3d_required_success(self):
+        self._test_3d('5212345678901234', success=True)
+
+    def test_create_transaction_visa(self):
+        self._test_card('5136333333333335')
+
+    def test_create_transaction_us(self):
+        self._test_card('4400000000000008')
+
+    def test_create_transaction_france(self):
+        self._test_card('4977949494949497')
+
+#    def test_create_transaction_risk_elevated(self):
+#        self._test_card(
+#            '4000000000009235',
+#            expected_risk_level='elevated')
+
+    def test_create_transaction_wrong_cvc(self):
+        with self.assertRaises(UserError):
+            self._test_card({
+                "number": '4977949494949497',
+                "expiryMonth": 10,
+                "expiryYear": 2020,
+                "cvc": 'wrong',
+                "holderName": 'John Doe',
+                }, expected_state='failed')
+
+    def test_create_transaction_wrong_number(self):
+        with self.assertRaises(UserError):
+            self._test_card({
+                "number": '497794fail',
+                "expiryMonth": 10,
+                "expiryYear": 2020,
+                "cvc": '737',
+                "holderName": 'John Doe',
+                }, expected_state='failed')
+
+    def test_create_transaction_wrong_expiry(self):
+        # NOTE that here Adyen returns a 000 error code
+        # which doesn't point to a wrong expiry as documented
+        # the message is correct but we decided not forward Adyen
+        # messages blindly, so the error is said to be unknown...
+        with self.assertRaises(UserError):
+            self._test_card({
+                "number": '4977949494949497',
+                "expiryMonth": 13,
+                "expiryYear": 2020,
+                "cvc": '737',
+                "holderName": 'John Doe',
+                }, expected_state='failed')
+
+    def test_wrong_api_key(self):
+        keychain = self.env['keychain.account']
+        account = keychain.sudo().retrieve([
+            ('namespace', '=', 'adyen')
+            ])[0]
+        account.write({'clear_password': 'wrong_api_key'})
+        with self.assertRaises(UserError):
+            self._test_card('5136333333333335')
+
```

Comment:
Can you add a test to check if the authentification failed?
You can modify the travis configuration to install the master version of adyen lib (that fix the issue when auth fail)

Reply:
well the failed auth will also re-raise a UserError (and that is what is tested here). So what do you suggest to test it further? Check the error message may be? Do you know a simple solution to check the error message? I didn't do that because I don't about any simple or standard OCA solution...